scipy/stats/_AB_SW.pyx

```diff
@@ -0,0 +1,463 @@
+    cdef double A1, A2, fac, delta, w1, y, ld, bf, gamma, m, s
+    cdef double RANGE, SA, SX, SSX, SSA, SAX, ASA, XSX, SSASSX, XX, XI
+    cdef double Z90F, Z95F, Z99F, ZFM, ZSD, ZBAR
+
+    if n1 < 0:
+        n1 = n
+    nn2 = n // 2
+    if nn2 < n2:
+        return 1., 1., 3
+    if n < 3:
+        return 1., 1., 1
+    w = 1.
+    pw = 1.
+    an = n
+
+
+    if not init:
+        if n == 3:
+            a[0] = SQRTH
+        else:
+            an25 = an + 0.25
+            summ2 = 0.
+            for ind1 in range(n2):
+                temp = _ppnd((ind1+1-0.375)/an25)
+                a[ind1] = temp
+                summ2 += temp**2
+
+            summ2 *= 2.
+            ssumm2 = sqrt(summ2)
+            rsn = 1 / sqrt(an)
+            A1 = _poly(c1, 6, rsn) - (a[0]/ssumm2)
+            if n > 5:
+                i1 = 2
+                A2 = -a[1]/ssumm2 + _poly(c2, 6, rsn)
+                fac = sqrt((summ2 - (2 * a[0]**2) - 2 * a[1]**2) / (1 - (2 * A1**2) - 2 * A2**2))
+                a[1] = A2
+            else:
+                i1 = 1
+                fac = sqrt((summ2 - 2 * a[0]**2)/(1 - 2 * A1**2))
+
+            a[0] = A1
+            for ind1 in range(i1, nn2):
+                a[ind1] *= -1./fac
+        init = True
+
+
+    if n1 < 3:
+        return w, pw, 1
+    ncens = n - n1
+
+    if ncens < 0 or ((ncens > 0) and (n < 20)):
+        return w, pw, 4
+    delta = ncens / an
+    if delta > 0.8:
+        return w, pw, 5
+
+
+    # Original Fortran code checked the W input value
+    # and acted on it via
+    # C
+    # C If W input as negative, calculate significance level of -W
+    # C
+    #       IF (W .LT. ZERO) THEN
+    #         W1 = ONE + W
+    #         IFAULT = 0
+    #         GOTO 70
+    #       END IF
+    # However W was marked as output in the fortran wrappers
+    # leading to undefined behavior of the uninitialized W
+    # Here W is assumed to be 0.
+    RANGE = x[n1-1] - x[0]
+    if RANGE < SMALL:
+        return w, pw, 6
+
+    XX = x[0] / RANGE
+    SX = XX
+    SA = -a[0]
+    ind2 = n - 2
+    for ind1 in range(1, n1):
+        XI = x[ind1] / RANGE
+        SX += XI
+        if ind1 != ind2:
+            SA += (-1 if ind1 < ind2 else 1)*a[min(ind1, ind2)]
+        XX = XI
+        ind2 -= 1
+
+    ifault = 0
+    if n > 5000:
+        ifault = 2
+
+    SA /= n1
+    SX /= n1
+    SSA, SSX, SAX = 0., 0., 0.
+    ind2 = n - 1
+    for ind1 in range(n1):
+        if ind1 != ind2:
+            ASA = (-1 if ind1 < ind2 else 1)*a[min(ind1, ind2)] - SA
+        else:
+            ASA = -SA
+
+        XSX = x[ind1]/RANGE - SX
+        SSA += ASA * ASA
+        SSX += XSX * XSX
+        SAX += ASA * XSX
+        ind2 -= 1
+
+    SSASSX = sqrt(SSA * SSX)
+    w1 = (SSASSX - SAX) * (SSASSX + SAX)/(SSA * SSX)
+    w = 1 - w1
+
+    # Calculate significance level for W (exact for N=3)
+    if n == 3:
+        pw = PI6 * (asin(sqrt(w)) - PI_OVER_3)
+        return w, pw, ifault
```

Comment:
In the Fortran version, this could result in negative p-values, so lines 1883-1886 of `_morestats.py` overrides the p-value for `n == 3`. If the Cython version here is more accurate, it's possible that those Python lines can be deleted. Alternatively, the formulation of the calculation in 1883-1886 can be moved here. (It's more obvious that the p-value cannot be negative in that formulation.)

Reply:
I checked the offending input in the reported issue and Cython version also returns negative numbers. Must be something algorithmic hence I adopted your fix directly baked into the Cython version with a note about the modification.